### PR TITLE
Integrate Tokens table data from the subgraph

### PR DESCRIPTION
**Pull request by ramirotw**
# Summary

image

### CODE REVIEW

**Comment by ramirotw** [comment]
@elena-zh FYI the Gnosis Chain tokens data is broken.  @GabrielCamba pushed the updated changes for the mainnet and gc subgraphs but only the gc subgraph finished indexing, I need to push the updates to fix the queries but I'll do it once the mainnet subgraph is done.

**Comment by elena-zh** [comment]
@ramirotw , I have left a few notes in the https://github.com/cowprotocol/explorer/pull/93#issuecomment-1132946911 PR, so the issues I found there are also related to the current PR. 

In addition, I have noticed that some data cannot be loaded here
image

And for 1 token I noticed '-infinity%' for rETH
image
 So it would be nice to change the percentage to '0' of '-' (dash) when it shows 'infinity'

Then, I noticed that 0 price can have different amount of decimals (Rinkeby)
image

Thanks!

File: src/components/token/TokenTable/index.tsx [hunk 1]
```diff
@@ -138,11 +138,11 @@ const HeaderTitle = styled.span`
 
 const TokenWrapper = styled.div`
   display: flex;
-  align-items: center;
   a {
-    display: none;
+    width: 100px;
```
Comment (alongoni) [inline src/components/token/TokenTable/index.tsx:142]: width: 10rem;

File: src/components/token/TokenTable/index.tsx [hunk 2]
```diff
@@ -138,11 +138,11 @@ const HeaderTitle = styled.span`
...
   }
   img {
+    margin-right: 10px;
```
Comment (alongoni) [inline src/components/token/TokenTable/index.tsx:145]: margin-right: 1rem;

hey @matextrem, we should use rem units instead of px. Thanks! :)

Comment (matextrem) [reply src/components/token/TokenTable/index.tsx:145]: Gotcha! Will update this

**Comment by ramirotw** [comment]
@alfetopito we have a problem:

image

Context: With the introduction of the `ipfs-only-hash` lib in https://github.com/cowprotocol/cow-sdk/commit/0708de8067d0560d1cfe66d0908b821ee81330bf#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R34 there's now an incompatibility of node versions as a dependency of `ipfs-only-hash` expects node to be >= 14. I have tested everything locally and the only thing I needed to bump because it was causing issues was `favicons-webpack-plugin`, the rest worked fine. Maybe this PR is a good moment to test moving to node 16.

**Comment by ramirotw** [comment]
1. @matextrem can you check why the tooltips are rendered outside the table?

image

2. Also, there's a runtime error when going to https://pr98--gpui.review.gnosisdev.com/address/0x1811be0994930fE9480eAEDe25165608B093ad7A

BTW: I fixed an issue with the 24h volume that was wrong in https://github.com/cowprotocol/explorer/pull/98/commits/89687bf91dbb99ab3362165b2eccc0ce62bc36b5

File: src/apps/explorer/components/TokensTableWidget/index.tsx [hunk 3]
```diff
@@ -156,3 +156,11 @@ export const TokensTableWidget: React.FC<Props> = () => {
     </TableWrapper>
   )
 }
+
+function undefinedToNull(n: number | undefined | null): number | null {
+  if (n === undefined) {
+    return null
+  }
+
+  return n
```
Comment (alfetopito) [inline src/apps/explorer/components/TokensTableWidget/index.tsx:165]: What about:

```suggestion
  return n ?? null
```

File: src/hooks/useGetTokens.ts [hunk 4]
```diff
@@ -244,7 +243,7 @@ export type TokenData = {
 }
 
 function getPercentageDifference(a: number, b: number): number {
-  return b ? ((a - b) / a) * 100 : 0
+  return a && b ? ((a - b) / a) * 100 : 0
```
Comment (alfetopito) [inline src/hooks/useGetTokens.ts:246]: If `b` is `0`, it'll return `0` as opposed to `100`

Is this the desired behaviour?

File: src/apps/explorer/components/TokensTableWidget/index.tsx (hunk 3 above)
Comment (alfetopito) [reply src/apps/explorer/components/TokensTableWidget/index.tsx:165]: You could actually replace the `|| null` with `?? null` and not need this function.

File: src/hooks/useGetTokens.ts (hunk 4 above)
Comment (ramirotw) [reply src/hooks/useGetTokens.ts:246]: if we do that then the % will be misleading, lets say the current (a) eth price is 2.000 but we don't have data for the last 7 days (b), would be ok to show 100% change on the Last 7 days column?

**Comment by henrypalacios** [comment]
I have a doubt here, when I was checking tx we can see in this PR as https://pr98--gpui.review.gnosisdev.com/tx/0xef86008c9da5a5dec0cfdc52a2d5134b1c8c25a13f2ce0d8b3eaa082befb2ed2?tab=orders
Selection_567


I see that txhash is not shown, Shouldn't we search like search bar does?

**Comment by ramirotw** [comment]
> @ramirotw I have a doubt here, when I was checking tx that we can see thanks to this PR 👇 https://pr98--gpui.review.gnosisdev.com/tx/0xef86008c9da5a5dec0cfdc52a2d5134b1c8c25a13f2ce0d8b3eaa082befb2ed2?tab=orders Selection_567
> 
> I see that txhash is not shown on the order details, Shouldn't we search like search bar does?

Good catch! The problem here is related to the prod/barn requests. When I did the changes to merge the responses the `/trades` endpoint was left behind, which is used here to get the tx hash. I'll push the fix to this branch to not complicate things.

File: src/hooks/useGetTokens.ts [hunk 5]
```diff
@@ -25,6 +34,78 @@ export function useGetTokens(networkId: Network | undefined): GetTokensResult {
...
+                value: Number(x.averagePrice),
+              }))
+              .sort((a, b) => a.time - b.time),
+          }
+
+          tokenData[address] = { lastDayUsdVolume, ...prices }
+        }
+
+        setTokens((tokens) => [...addHistoricalData(tokens, tokenData)])
+        setHistoricalDataLoaded((loaded) => ({ ...loaded, [pageIndex]: true }))
```
Comment (alfetopito) [inline src/hooks/useGetTokens.ts:97]: Could some sections of this function be extracted outside the hook body?
Feels like too much is going on in here and the logic could be abstracted away

File: src/hooks/useGetTokens.ts (hunk 4 above)
Comment (alfetopito) [reply src/hooks/useGetTokens.ts:250]: Then you don't need that `if` on top, and can remove the requirement for `b` to be truthy:

```suggestion
  return a ? ((a - b) / a) * 100 : 0
```

If `b` is `0`, a/a * 100 => 100

**Review by alfetopito — APPROVED** [review]
Overall quite happy with the current state.

Left 2 nitpick comments that are not blocking.

Only left concern is regarding the data, which @GabrielCamba is taking care of

Interested in getting @elena-zh 's feedback as well

**Comment by elena-zh** [comment]
@ramirotw , great job!
However, here are issues I have found: 
1. It would be nice to reset the table's pagination when change networks (to reproduce navigate to the 2 page and change a network in the selector: the table remains to show the 2nd page)
2. I wonder what's the difference between '0%' and '-' in the table's values? 
image

4. I still can see different decimals precision for 0 prices: 

5. Token name is shifted in a mobile view
name mobile
6. # is shifted in a mobile view
align
7. Again, what is the difference in the 'dash' and in the horizontal graph if everywhere 0-s are displayed? Also, I see graphs have a different size
difference
8. I see this vertical line on the graph. Should it show something? 
what shoud show

9. When a token does not have a name, maybe we should show its symbol instead, WDYT?
no name = symbol
10. Seems like WETH (or ETH?) does not have a symbol/name in Rinkeby
weth
11. Could we please add a tooltip with a full amount to volumes?
tooltips for volumes

Thanks!

**Comment by elena-zh** [comment]
11. When search for a token, and search shows more than 1 page, it would be nice to show an actual number of pages with search results instead of all 5 pages. See the video: https://watch.screencastify.com/v/k8qepZsT9uDi8xWc7qv8

**Comment by ramirotw** [comment]
Thanks @elena-zh for the feedback!


The dash is the absence of data from the subgraph. 
> 2\. I wonder what's the difference between '0%' and '-' in the table's values?
>     image

Same as above, the dash is the absence of data. The chart uses 7 data points max, but if we have less data points than that, we still show the chart, hence different charts with different sizes.
> 6\. Again, what is the difference in the 'dash' and in the horizontal graph if everywhere 0-s are displayed? Also, I see graphs have a different size
>     difference

File: src/hooks/useGetTokens.ts (hunk 4 above)
Comment (ramirotw) [reply src/hooks/useGetTokens.ts:250]: 🤦‍♂️this is what happens when you don't sleep enough. Thanks

**Comment by elena-zh** [comment]
Hey @ramirotw , great job!
I have created a separate enhancement for the scroll to top action in a mobile view #117 . 

Also, it would be great to fix 'Total volume' column name in a desktop view ('volume' should start with a lowercase letter)
image

**Comment by elena-zh** [comment]
I have also found an issue that is related to the difference in data in BARN and PROD APIs. I assume, end users can face it, so I reported it: #118 
@ramirotw , @alfetopito , please review it and let me know if we need to fix  or close it. Thanks

**Review by elena-zh — APPROVED** [review]
Approved now!

**Pull request merged**